### ISSUE TITLE
A few questions and comments

### ISSUE THREAD

**Issue by FourierTransformer** [teal-language/tl#317]
So, I was playing around with tl for ftcsv (in [this branch](https://github.com/FourierTransformer/ftcsv/tree/convert-to-tl)) and it is really cool, and pretty easy to use. I'm glad to see a minimalistic type system for lua! Congrats to the team behind it! I even found a bug because of it!

1) I did have a couple of questions about using tl. I have this basic file loading snippet, and depending on the ftcsv options will either read in the whole file (with `*alll`) or just a certain number of bytes. So I went with the type definition of `string | number` but noticed I still needed to typecheck to use it in `file:read()` even though `read()` can take in either. Am I doing something wrong with the union or is this the way it is?
```lua
local function loadFile(textFile: string, amount: string | number)
    local file = io.open(textFile, "r")
    if not file then error("ftcsv: File not found at " .. textFile) end
    local lines: string
    if amount is string then
        file:read(amount)
    else
        file:read(amount)
    end
    if amount == "*all" then
        file:close()
    end
    return lines, file
end
```

2) The other thing I ran into was using `{{string:any}}`. When trying to encode a csv from a lua table that looks something like this: `a= {{b="hey", c=true, d=43}, {b="yo", c=false, d=42}}`. In some of the encoder code, I was trying to use `{{string:any}}` but ran into the error below. I worked around it by creating a `local type CSVRow = {string: any}` and switching form `{{string:any}}` to `{CSVRow}`. I'm mostly curious if that's the intended way to fix that issue and if I contributed types for ftcsv in the type definitions project would people need to cast their tables as `CSVRow` to make things work?
```
5 errors:
encoder.lua:90:5: in assignment: got {{string : <any type>}}, expected {{string : <any type>}}
encoder.lua:152:43: argument 1: got {{string : <any type>}}, expected {{string : <any type>}}
encoder.lua:154:30: argument 2: got {{string : <any type>}}, expected {{string : <any type>}}
encoder.lua:163:49: argument 1: got {{string : <any type>}}, expected {{string : <any type>}}
encoder.lua:165:37: argument 1: got {{string : <any type>}}, expected {{string : <any type>}}
```

3) Can type definitions be auto-generatred from tl files? For my stuff it's not a lot to write, but would be cool for other stuff!

I can create separate issues if needed. So far in my time with tl, it's been pretty amazing!

**Comment by hishamhm** [comment]
> I even found a bug because of it!

Hi, thank you for the positive feedback!! This is amazing to hear!

> Am I doing something wrong with the union or is this the way it is?

You're not! The problem was that `file:read` was declared in the internal standard library definitions as a "polymorphic" function (in other words, there were two separate definitions for `read`, one with string and another with number, and they were tried one by one). I will update the standard library definition to use the union as a short-term fix. Making the type system smart enough to decide when a union value is safe to use for polymorphic functions would be a better fix but it's a lot more work; I'm going to add this as a separate issue.

The short-term fix should be here in #320 (with it, the redundant `if amount is string` check shouldn't be needed)

> 2. The other thing I ran into was using {{string:any}}.

I believe this is fixed by #319.

**Comment by hishamhm** [comment]
> Can type definitions be auto-generatred from tl files? For my stuff it's not a lot to write, but would be cool for other stuff!

Not sure what you mean by that, could you clarify?

**Comment by FourierTransformer** [comment]
> Not sure what you mean by that, could you clarify?

Yeah! So, eventually I'll have an `ftcsv.tl` file that will be transpiled into a `ftcsv.lua` file. It would be nice if it could optionally output a `ftcsv.d.tl` for the public methods in that file. Effectively just auto-generating the definitions for the `parse`, `parseLine`, and `encode` methods that are returned when importing the library.

I'll try out some of the branches in a bit and let you know how it goes! Thanks for the quick response!

**Comment by lenscas** [comment]
I'm guessing that this is for distributing a lua library written in teal? TS does something similar I believe.

**Comment by hishamhm** [comment]
The only use of `.d.tl` files at this time is to make Teal code interface with Lua. If the code is already written in Teal, shipping the `.tl` and `.lua` files should be sufficient, I guess?

**Comment by lenscas** [comment]
but right now you would need to manually merge the build output and your codebase, don't you? Or can tell teal to output in the same directory as your code and not have it mess up?

**Comment by darrenjennings** [comment]
FWIW in typescript projects, the tsconfig (and ts version) varies from project to project so it is not correct to distribute the ts source and ask the users to compile it themselves, but instead library authors distribute the js, types and source maps. Extracting the types is part of tsc output (if desired via `--declaration`)

**Comment by FourierTransformer** [comment]
> If the code is already written in Teal, shipping the .tl and .lua files should be sufficient, I guess?

Maybe, if LuaRocks supported teal? Or if teal had its own package manager?

**Comment by lenscas** [comment]
now, although a custom package manager could do nice things (especially if stuff like https://github.com/teal-language/tl/pull/301 gets integrated in it) I don't think a custom package manager is a good thing.

no need to split a userbase up like that, especially not one that isn't super big already...

**Comment by hishamhm** [comment]
Both fixes were merged (and included in 0.10.1), so I'm closing thes one!
